sql/src/main/java/io/crate/planner/operators/Join.java

```diff
@@ -0,0 +1,328 @@
+            // TODO: probably can be applied earlier - might only match a single relation but isn't pushed down due to outer join?
+            if (!queryParts.isEmpty()) {
+                join = new Filter(join, new WhereClause(AndOperator.join(queryParts.values())));
+            }
+            return join;
+        };
+    }
+
+    @Nullable
+    private static <V> V removeMatch(Map<Set<QualifiedName>, V> valuesByNames, Set<QualifiedName> names, QualifiedName nextName) {
+        for (QualifiedName name : names) {
+            V v = valuesByNames.remove(Sets.newHashSet(name, nextName));
+            if (v != null) {
+                return v;
+            }
+        }
+        return null;
+    }
+
+    private static void addColumnsFrom(Iterable<? extends Symbol> symbols,
+                                       Consumer<? super Symbol> consumer,
+                                       QueriedRelation rel) {
+
+        for (Symbol symbol : symbols) {
+            addColumnsFrom(symbol, consumer, rel);
+        }
+    }
+
+    private static void addColumnsFrom(@Nullable Symbol symbol, Consumer<? super Symbol> consumer, QueriedRelation rel) {
+        if (symbol == null) {
+            return;
+        }
+        FieldsVisitor.visitFields(symbol, f -> {
+            if (f.relation().getQualifiedName().equals(rel.getQualifiedName())) {
+                consumer.accept(rel.querySpec().outputs().get(f.index()));
+            }
+        });
+    }
+
+    private static Map<Set<QualifiedName>,Symbol> getQueryParts(WhereClause where) {
+        if (where.hasQuery()) {
+            return QuerySplitter.split(where.query());
+        }
+        return Collections.emptyMap();
+    }
+
+    private Join(LogicalPlan lhs, LogicalPlan rhs, JoinType joinType, @Nullable Symbol joinCondition) {
+        this.lhs = lhs;
+        this.rhs = rhs;
+        this.joinType = joinType;
+        this.joinCondition = joinCondition;
+        if (joinType == JoinType.SEMI) {
+            this.outputs = lhs.outputs();
+        } else {
+            this.outputs = Lists2.concat(lhs.outputs(), rhs.outputs());
+        }
+        this.baseTables = new ArrayList<>();
+        this.baseTables.addAll(lhs.baseTables());
+        this.baseTables.addAll(rhs.baseTables());
+        this.expressionMapping = new HashMap<>();
+        this.expressionMapping.putAll(lhs.expressionMapping());
+        this.expressionMapping.putAll(rhs.expressionMapping());
+    }
+
+    @Override
+    public Plan build(Planner.Context plannerContext,
+                      ProjectionBuilder projectionBuilder,
+                      int limit,
+                      int offset,
+                      @Nullable OrderBy order,
+                      @Nullable Integer pageSizeHint) {
+        Plan left = lhs.build(plannerContext, projectionBuilder, NO_LIMIT, 0, null, null);
+        Plan right = rhs.build(plannerContext, projectionBuilder, NO_LIMIT, 0, null, null);
+
+        // TODO: distribution planning
```

Comment:
Did we gain any performance from the distributed planning?

Reply:
I've had a few cases where it was actually a lot slower. But it depends on the join conditions and how many matches there are.

We should definetly invest some effort at some point to come up with a better heuristic to decide if it should run the nestedloop distributed as well or not.

But to get this into master I'd like to keep the behaviour as close to before as possible - so I marked it as TODO